### PR TITLE
Example: Generic Cleanup

### PR DESCRIPTION
**Pull request by addisoncrump**
Trying #2283 again, but more progressively.

### CODE REVIEW

File: libafl/src/events/mod.rs [hunk 1]
```diff
@@ -452,13 +448,26 @@ pub trait EventFirer: UsesState {
 }
 
 /// [`ProgressReporter`] report progress to the broker.
-pub trait ProgressReporter: EventFirer
-where
-    Self::State: HasMetadata + HasExecutions + HasLastReportTime,
-{
```
Comment (addisoncrump) [inline libafl/src/events/mod.rs:458]: This is the kind of change I was talking about, @domenukk. We can just straight up avoid all these generics here and shift them to impls.

File: libafl/src/events/mod.rs [hunk 2]
```diff
@@ -452,13 +448,26 @@ pub trait EventFirer: UsesState {
...
+        state: &mut Self::State,
+        monitor_timeout: Duration,
+    ) -> Result<(), Error>;
+
+    /// Send off an info/monitor/heartbeat message to the broker.
+    /// Will return an [`Error`], if the stats could not be sent.
+    fn report_progress(&mut self, state: &mut Self::State) -> Result<(), Error>;
+}
+
+impl<EF> ProgressReporter for EF
```
Comment (domenukk) [inline libafl/src/events/mod.rs:466]: Doesn't this mean downstream ProgressReporters cannot implement their own behavior anymore?

Comment (addisoncrump) [reply libafl/src/events/mod.rs:466]: Sure, but every progress reporter has used the default impls. If we want, we can also split this into `impl DefaultProgressReporter` or even just generic functions, e.g. `default_report_progress`

Comment (domenukk) [reply libafl/src/events/mod.rs:466]: Point is, right now overriding some default behavior "just works" after this it won't work anymore

Comment (domenukk) [reply libafl/src/events/mod.rs:466]: For a library user this makes a difference. And if we want to add new methods to a trait it matters as well

Comment (addisoncrump) [reply libafl/src/events/mod.rs:466]: Certainly. We have to make these decisions in an opinionated manner, on a case-by-case basis.

From my perspective, there is very rarely any reason to actually implement alternative progress reporters. We have two real options to reduce generics:
 - Any option we choose that doesn't enforce a specific implementation would remove generics from ~everything trying to write anything involving a progress reporter at the cost of ~10 lines of code per implementation of progress reporter.
 - We enforce a specific implementation for everyone. We invalidate downstream implementions of ProgressReporter, but make the LoC cost for all implementations 0.

**Pull request closed without merging**